File: question/leetcode/leetcode1019/leetcode1019.go
package leetcode

import "clarencedu/algorithm-go/structure"

type ListNode = structure.ListNode

// INFO: 寻找下一个更大的元素，经典的单调栈问题
// 和数组的单调栈问题类似，只不过这里是链表,外层循环遍历变成了移动链表指针遍历
func nextLargerNodes(head *ListNode) []int {
	var res []int
	var stack [][]int //存储元素值和索引,[[val,index],[val,index]]
	cur := head
	idx := -1

	for ; cur != nil; cur = cur.Next {
		idx++
		// 初始化res，占位，后面会被替换
		res = append(res, 0)
		// 找到比当前元素大的元素
		for len(stack) > 0 && stack[len(stack)-1][0] < cur.Val {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			//这一题只需要保存值，不需要保存索引
			//否则 res[top[1]] = idx
			res[top[1]] = cur.Val
		}
		stack = append(stack, []int{cur.Val, idx})
	}

	return res
}

File: question/leetcode/leetcode1019/leetcode1019.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def nextLargerNodes(head: Optional[ListNode]) -> List[int]:
    res = []
    stack = []  # 存储 (值, 索引) 元组
    cur = head
    idx = -1

    while cur:
        idx += 1
        res.append(0)  # 预占位
        # 维护单调栈，栈顶元素小于当前值时出栈
        while stack and stack[-1][0] < cur.val:
            val, i = stack.pop()
            res[i] = cur.val  # 赋值下一个更大的值
        stack.append((cur.val, idx))
        cur = cur.next

    return res